python_modules/dagster/dagster/core/errors.py

```diff
@@ -138,6 +147,31 @@ class DagsterPY4JTribalKnowledgeException(Exception):
     '''
 
 
+@contextmanager
+def user_code_error_boundary(error_cls, msg, **kwargs):
+    '''
+    Wraps the execution of user-space code in an error boundary. This places a uniform
+    policy around an user code invoked by the framework. This ensures that all user
+    errors are wrapped in the DagsterUserCodeExecutionError, and that the original stack
+    trace of the user error is preserved, so that it can be reported without confusing
+    framework code in the stack trace, if a tool author wishes to do so. This has
+    been especially help in a notebooking context.
+    '''
+    check.str_param(msg, 'msg')
+    check.subclass_param(error_cls, 'error_cls', DagsterUserCodeExecutionError)
+
+    try:
+        yield
+    except Exception as e:  # pylint: disable=W0703
+        if isinstance(e, DagsterError):
```

Comment:
let's add a comment about what is going on here.

in case 1) the system has thrown an error or an error that is part of the user-framework contract has been thrown (e.g. type checking error)

in case 2) there has been an error indicated that something has gone wrong and computation should cease, and an error reported somewhere up the stack